perun-core/src/main/java/cz/metacentrum/perun/core/impl/AuthzResolverImpl.java

```diff
@@ -146,21 +115,27 @@ public AuthzResolverImpl(DataSource perunPool) {
 	}
 
 	@Override
-	public AuthzRoles getRoles(User user) {
+	public AuthzRoles getRoles(User user, boolean getAuthorizedGroupBasedRoles) {
 		AuthzRoles authzRoles = new AuthzRoles();
 
 		if (user != null) {
 			try {
-				// Get roles from Authz table
-				List<Pair<String, Map<String, Set<Integer>>>> authzRolesPairs = jdbc.query("select " + authzRoleMappingSelectQuery
-						+ ", roles.name as role_name from authz left join roles on authz.role_id=roles.id where authz.user_id=? or authorized_group_id in "
-						+ "(select groups.id from groups join groups_members on groups.id=groups_members.group_id and groups_members.source_group_status=? join members on "
-						+ "members.id=groups_members.member_id join users on users.id=members.user_id where users.id=? and members.status=?)", AUTHZROLE_MAPPER, user.getId(), MemberGroupStatus.VALID.getCode(), user.getId(), Status.VALID.getCode());
+				// Get directly assigned user roles from Authz table
+				List<Pair<String, Map<String, Set<Integer>>>> authzRolesPairs = jdbc.query(
+					"select " + authzRoleMappingSelectQuery	+ ", roles.name as role_name " +
+						"from authz left join roles on authz.role_id=roles.id " +
+						"where authz.user_id=?",
+					AUTHZROLE_MAPPER, user.getId());
 
 				for (Pair<String, Map<String, Set<Integer>>> pair : authzRolesPairs) {
 					authzRoles.putAuthzRoles(pair.getLeft(), pair.getRight());
 				}
 
+				if (getAuthorizedGroupBasedRoles) {
+					// add user roles based on membership in authorized groups
+					authzRoles.putAll(getRolesObtainedFromAuthorizedGroupMemberships(user));
```

Comment:
I think this removes existing values for keys (roles) and puts new ids to them (so objects are not merged, but overwritten with group-gained roles).

Reply:
Very nice! Previous tests were not built to test both direct and inherited roles presence. I updated them and fixed the problem with removing already existing roles. Thank you.